File: linux_admin/process_states/zombie.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

int main() {
    pid_t pid = fork();

    if (pid < 0) {
        perror("fork");
        exit(EXIT_FAILURE);
    }

    if (pid == 0) {
        // Child process
        printf("Child process (PID: %d) is exiting\n", getpid());
        exit(EXIT_SUCCESS);
    } else {
        // Parent process
        printf("Parent process (PID: %d) sleeping\n", getpid());
        sleep(10); // Child will be a zombie during this time
        printf("Parent process (PID: %d) waking up\n", getpid());
        wait(NULL); // Collect child process exit status
        printf("Zombie child process has been reaped\n");
    }
    return 0;
}

File: linux_admin/process_states/zombie.py
import os
import time

def main():
    pid = os.fork()

    if pid > 0:
        # Parent process
        print(f"Parent process (PID: {os.getpid()}) sleeping")
        time.sleep(10)  # Child will be a zombie during this time
        print(f"Parent process (PID: {os.getpid()}) waking up")
        os.wait()  # Collect child process exit status
        print("Zombie child process has been reaped")
    else:
        # Child process
        print(f"Child process (PID: {os.getpid()}) is exiting")
        os._exit(0)

if __name__ == "__main__":
    main()
